### PR TITLE
Validate User authorization before sync

### PR DESCRIPTION
**Pull request by githengi**
PR for Issue [https://github.com/OpenSRP/opensrp-client-core/issues/156](https://github.com/OpenSRP/opensrp-client-core/issues/156)

Fixes [https://github.com/OpenSRP/opensrp-client-core/issues/163](https://github.com/OpenSRP/opensrp-client-core/issues/163)

### CODE REVIEW

File: opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java [hunk 1]
```diff
@@ -78,6 +86,45 @@ private void doSync() {
         }
     }
 
+    private void verifyAuthorization() {
+        org.smartregister.Context opensrpContent = CoreLibrary.getInstance().context();
+        String baseUrl = opensrpContent.configuration().dristhiBaseURL();
+        if (baseUrl.endsWith("/")) {
+            baseUrl = baseUrl.substring(0, baseUrl.lastIndexOf("/"));
```
Comment (vincent-karuri) [inline opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: Why not use  `baseUrl.substring(0, baseUrl.length - 1);`?

Currently, it searches the string again for the last occurence of `/`, right?

File: opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java [hunk 2]
```diff
@@ -78,6 +86,45 @@ private void doSync() {
...
+            int statusCode = urlConnection.getResponseCode();
+            Log.i(TAG, urlConnection.getResponseMessage());
+            urlConnection.disconnect();
+            if (statusCode == HttpStatus.SC_UNAUTHORIZED) {
+                Log.i(TAG, "User not authorized. User access was revoked, will log off user");
+                opensrpContent.userService().forceRemoteLogin();
+                Intent intent = new Intent(Intent.ACTION_MAIN);
+                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                intent.addCategory(Intent.CATEGORY_HOME);
+                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
```
Comment (vincent-karuri) [inline opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: Does this need to be set twice?

File: opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    private Context context;
+
+    public SyncUtils(Context context) {
+        this.context = context;
+    }
+
+    public boolean verifyAuthorization() {
+        String baseUrl = opensrpContent.configuration().dristhiBaseURL();
+        if (baseUrl.endsWith("/")) {
+            baseUrl = baseUrl.substring(0, baseUrl.lastIndexOf("/"));
```
Comment (vincent-karuri) [inline opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java]: Why not use baseUrl.substring(0, baseUrl.length - 1);?

Currently, it searches the string again for the last occurence of /, right?

File: opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        final String username = opensrpContent.allSharedPreferences().fetchRegisteredANM();
+        final String password = opensrpContent.allSettings().fetchANMPassword();
+        baseUrl = baseUrl + "/user-details?anm-id=" + username;
```
Comment (vincent-karuri) [inline opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java]: Add `/user-details?anm-id="` as a constant.

File: opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        Intent intent = new Intent(Intent.ACTION_MAIN);
+        intent.addCategory(Intent.CATEGORY_LAUNCHER);
+        intent.setPackage(context.getPackageName());
+        //retrieve the main/launcher activity defined in the manifest and open it
+        List<ResolveInfo> activities = context.getPackageManager().queryIntentActivities(intent, 0);
+        if (activities.size() == 1) {
+            intent = intent.setClassName(context.getPackageName(), activities.get(0).activityInfo.name);
+            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+            intent.addCategory(Intent.CATEGORY_HOME);
+            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
```
[the remark this answers is not in the document]
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java]: I remove this

File: opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java [hunk 6]
```diff
@@ -69,15 +72,24 @@ private void doSync() {
         }
 
         try {
-            pushToServer();
-            pullECFromServer();
+            boolean hasValidAuthorization = syncUtils.verifyAuthorization();
+            if (hasValidAuthorization ||
+                    (!hasValidAuthorization && !CoreLibrary.getInstance().getSyncConfiguration().disableSyncToServerIfUserIsDisabled())) {
```
Comment (vincent-karuri) [inline opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: Remove this `!hasValidAuthorization`.

**Pull request merged**